component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mobileconnect/MobileConnectAuthenticator.java

```diff
@@ -1082,26 +1085,53 @@ private HttpURLConnection msisdnBasedDiscoveryProcess(String authorizationHeader
     private void buildClaims(AuthenticationContext context, String jsonString)
             throws ApplicationAuthenticatorException {
 
-        Map<String, Object> userClaims;
-        userClaims = JSONUtils.parseJSON(jsonString);
+        final String claimDialectUri = getClaimDialectURI();
+        final boolean shouldPrefixClaimDialectUri = shouldPrefixClaimDialectUri();
+        final String userClaimUri = context.getExternalIdP().getIdentityProvider().getClaimConfig().getUserClaimURI();
+        Map<String, Object> userClaims = JSONUtils.parseJSON(jsonString);
+
         if (userClaims != null) {
             Map<ClaimMapping, String> claims = new HashMap<>();
+            String claimUri, claimValue;
+
             for (Map.Entry<String, Object> entry : userClaims.entrySet()) {
-                claims.put(ClaimMapping.build(entry.getKey(), entry.getKey(), null,
-                        false), entry.getValue().toString());
+                claimUri = getEffectiveClaimDialectUri(claimDialectUri, entry.getKey(),
+                        shouldPrefixClaimDialectUri);
+                claimValue = entry.getValue().toString();
+
+                claims.put(ClaimMapping.build(claimUri, claimUri, null,
+                        false), claimValue);
+
                 if (log.isDebugEnabled() &&
                         IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.USER_CLAIMS)) {
-                    log.debug("Adding claim mapping : " + entry.getKey() + " <> " + entry.getKey() + " : "
-                            + entry.getValue());
+                    log.debug("Adding claim mapping : " + claimUri + " <> " + claimUri + " : "
+                            + claimValue);
                 }
             }
-            if (StringUtils.isBlank(context.getExternalIdP().getIdentityProvider().getClaimConfig().
-                    getUserClaimURI())) {
+
+            if (StringUtils.isBlank(userClaimUri)) {
+                //Set id as the default userClaimUri in IDP claim Configs.
                 context.getExternalIdP().getIdentityProvider().getClaimConfig().setUserClaimURI
-                        (MCAuthenticatorConstants.CLAIM_ID);
+                        (getEffectiveClaimDialectUri(claimDialectUri,
+                                MCAuthenticatorConstants.DEFAULT_USER_IDENTIFIER, shouldPrefixClaimDialectUri));
+            }
+
+            String subjectFromClaims = null;
+            if (StringUtils.isNotBlank(claimDialectUri)) {
+                setSubject(context, userClaims);
+                context.getSubject().setUserAttributes(claims);
```

Comment:
Hasintha has pointed out since claims are setting out in here, we don't need to do that at line 1142. Instead we can use context.getSubject().setAuthenticatedSubjectIdentifier(subjectFromClaims) if subject is not null in there.